drivers: sensor: sht3xd: migrate from Kconfig to device tree and add sample

pabigot opened the pull request:
There's at least one gap here, in that the Kconfig to enable `SHT3XD_TRIGGER` should probably be conditional on a non-empty `DT_SHT3XD_GPIO_ALERT_DEV_NAME`.  I don't know how to make that happen, and can't find any exemplars to use as a model.

Also I'm not sure the sample is properly integrated into the test infrastructure (it's executed, but no boards are configured for it).  I tested with a `nrf51_ble400`, but could add it to any DK-class Nordic board if desired.

pabigot committed:
drivers: sensor: sht3xd: migrate I2C from Kconfig to device tree

Replace Kconfig I2C bus and address information with device tree
bindings.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: sensor: sht3xd: migrate GPIO from Kconfig to device tree

Replace Kconfig ALERT signal GPIO device and pin information with device
tree bindings.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
samples: sensor: add sample for SHT3XD sensor

Added sample code for Sensirion SHT3XD humidity sensor.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

ulfalizer commented:
> There's at least one gap here, in that the Kconfig to enable `SHT3XD_TRIGGER` should probably be conditional on a non-empty `DT_SHT3XD_GPIO_ALERT_DEV_NAME`. I don't know how to make that happen, and can't find any exemplars to use as a model.

It's not possible to have Kconfig symbols depend on information from DT at the moment, but it's planned. @galak is working on making DT run before Kconfig, which needs to happen first.

pabigot commented:
@ulfalizer Thanks; that's what I remembered, but couldn't reconstruct where I'd seen that mentioned and how close it was to being merged.  I'm working on a bunch of sensor addition/extensions where I'd like to use device tree; if Zephyr won't be ready for a while I'll hold on to the rest of the patches.

galak commented:
> @ulfalizer Thanks; that's what I remembered, but couldn't reconstruct where I'd seen that mentioned and how close it was to being merged. I'm working on a bunch of sensor addition/extensions where I'd like to use device tree; if Zephyr won't be ready for a while I'll hold on to the rest of the patches.

Why not just remove the Kconfig SHT3XD_TRIGGER symbol and do something like if defined(DT_SHT3XD_GPIO_ALERT_DEV_NAME) in the code?

dbkinder commented on samples/sensor/sht3xd/README.rst:
```diff
@@ -0,0 +1,52 @@
+References
+**********
+
+ - SHT3X-DIS: https://www.sensirion.com/en/environmental-sensors/humidity-sensors/digital-humidity-sensors-for-various-applications/
```
```suggestion
 - `SHT3X sensor <https://www.sensirion.com/en/environmental-sensors/humidity-sensors/digital-humidity-sensors-for-various-applications/>`_
```

pabigot replied:
Should be fixed in latest push.

pabigot commented:
@galak My thought was that if the hardware doesn't have the signals required to support triggers then the application/user should be prevented from enabling it in Kconfig.  `TRIGGER_NONE` is the only satisfiable configuration in that situation.  Right now it's not clear there's a way to enforce that.

dbkinder reviewed (changes requested):
one tweak to the doc

galak commented:
> @galak My thought was that if the hardware doesn't have the signals required to support triggers then the application/user should be prevented from enabling it in Kconfig. `TRIGGER_NONE` is the only satisfiable configuration in that situation. Right now it's not clear there's a way to enforce that.
> 
> (The application choice of global or own thread triggers, and for some sensors the type of trigger, are still things that I think belong in Kconfig.)

I guess are you wanting apps to be able to do something like:

```
#ifdef CONFIG_SHT3XD_TRIGGER_NONE
...
#else 
...
#endif 
```

Just trying to understand why we need the Kconfig option for `TRIGGER` at all.

pabigot commented:
That the signal is connected in hardware allows trigger to be supported, but the application might not need to use it.  Further, the trigger implementation may be configurable, as with:
```
SHT3XD_TRIGGER_GLOBAL_THREAD
SHT3XD_TRIGGER_OWN_THREAD
SHT3XD_THREAD_PRIORITY
SHT3XD_THREAD_STACK_SIZE
```

All these are referenced only in the driver source, and none are meaningful if `alert-gpios` is missing from the device tree (i.e. not connected in hardware).  Are these not things that should be controlled through Kconfig?  If they are, then IMO exposing the UI to set them should be conditioned on information from the device tree.

Beyond triggers I understood there was agreement that while things like signal active level (high/low) might be in the device tree, whether the interrupt was edge or level triggered is something that would remain in Kconfig (also [high/low accuracy edge detection](https://github.com/zephyrproject-rtos/zephyr/issues/10339#issuecomment-440148175)).  Those also should be settable only when the signal is connected in hardware; when it isn't connected unsatisfiable settings in `prj.conf` should produce diagnostics.

If there's another way than Kconfig to get application configuration decisions propagated into driver code then maybe Kconfig doesn't need information from the device tree, but I thought that was its role.

pabigot commented:
@galak I might have finally understood your point once I noticed the leftover `SHT3XD_TRIGGER`.  I've added two commits that remove it and allow `SHT3XD_TRIGGER_NONE` to substitute, which is what I had in mind.  Is this approach better? worse? ...?

galak commented:
> @galak I might have finally understood your point once I noticed the leftover `SHT3XD_TRIGGER`. I've added two commits that remove it and allow `SHT3XD_TRIGGER_NONE` to substitute, which is what I had in mind. Is this approach better? worse? ...?

I think its an improvement, however I think we should be dealing with such things dynamically not via Kconfig.  The reason I say that, is what happens if for some reason you have 2 sensors in a system one with trigger and one w/o.  How would you configure or deal with that?

pabigot commented:
@galak Haven't thought about that yet.  I don't see much demand for two instances of the same sensor on the same device, as most sensors are measuring something so localized having two on the device wouldn't make any sense.  An SDP810 (Zephyr doesn't support yet) might usefully have multiple instances; same with thermistors.

In any case, I don't believe this discussion is specific to the changes in the PR and it's certainly taking things in a direction I haven't seen discussed before.  Do your concerns block this PR?  Can this discussion be moved to a new issue?

galak commented:
> @galak Haven't thought about that yet. I don't see much demand for two instances of the same sensor on the same device, as most sensors are measuring something so localized having two on the device wouldn't make any sense. An SDP810 (Zephyr doesn't support yet) might usefully have multiple instances; same with thermistors.
> 
> In any case, I don't believe this discussion is specific to the changes in the PR and it's certainly taking things in a direction I haven't seen discussed before. Do your concerns block this PR? Can this discussion be moved to a new issue?

This concern doesn't block the PR, just wanted to raise it to your attention as you think or look at other sensors.

galak reviewed (dismissed):
I'd personally keep CONFIG_SHT3XD_TRIGGER.  I think the ifndef is harder to read, and the one in Cmake is pretty subtle.  Having the extra "hidden" Kconfig symbol seems cleaner from a readability point of view to me.

pabigot commented:
@galak OK, so it's not an improvement ;-)

I'm going to wait a day or two for other comments; if there are no other changes or concerns I'll drop the patches that removed the option and do another test/push round.

galak commented:
> @galak OK, so it's not an improvement ;-)
> 
> I'm going to wait a day or two for other comments; if there are no other changes or concerns I'll drop the patches that removed the option and do another test/push round.

ok, once you drop the patches I'll approve this.

pabigot commented:
@galak dropped the patches as requested.

dbkinder commented on samples/sensor/sht3xd/README.rst:
```diff
@@ -0,0 +1,52 @@
+References
+**********
+
+ - `SHT3X-DIS sensor <https://www.sensirion.com/en/environmental-sensors/humidity-sensors/digital-humidity-sensors-for-various-applications/>`
```
Sorry, my bad, there needs to be a trailing underscore:

```suggestion
 - `SHT3X-DIS sensor <https://www.sensirion.com/en/environmental-sensors/humidity-sensors/digital-humidity-sensors-for-various-applications/>`_
```

pabigot replied:
No, my bad; I missed that when transcribing the change (it's years since I used restructured text).  Is it fixed now?

dbkinder reviewed (approved):
+1 for docs

MaureenHelm commented on samples/sensor/sht3xd/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.8.2)
```
Please add a license/copyright header.

@nashif Should the new github checks be catching this now?

MaureenHelm commented on samples/sensor/sht3xd/prj.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_STDOUT_CONSOLE=y
```
Please add a license/copyright header

MaureenHelm commented on samples/sensor/sht3xd/src/main.c:
```diff
@@ -0,0 +1,96 @@
+
+		rc = sensor_sample_fetch(dev);
+		if (rc == 0) {
+			rc = sensor_channel_get(dev, SENSOR_CHAN_AMBIENT_TEMP, &temp);
```
It doesn't make sense to fetch and get the sensor data in the trigger handler, then do it again here. Can you refactor such that the trigger handler just gives a semaphore?

pabigot replied:
It got a bit complex but I've replaced it with something a little more realistic.  All issues should be resolved; let me know if you want something more.

pabigot committed:
drivers: sensor: sht3xd: correct humidity threshold conversion

Original scale factor assumed a maximum humidity of 10^5 percent.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

MaureenHelm commented on samples/sensor/sht3xd/dts_fixup.h:
```diff
@@ -0,0 +1,13 @@
+
+#define DT_SHT3XD_NAME ""
+#define DT_SHT3XD_I2C_ADDR 0
+#define DT_SHT3XD_I2C_MASTER_DEV_NAME ""
```
I'd like to avoid this pattern for samples as it will cause users to think this sample will work on their board automatically, when in reality they need to create a dts overlay to define how the sensor is connected to the mcu. With the fixup, the compile will succeed but the application will fail at runtime. We should instead limit CI to build this sample only for board(s) that support it.

You also need to add a sample.yaml file to tell CI which boards and configurations to build the sample for. See `samples/sensor/fxos8700/sample.yaml` for an example.

pabigot replied:
Good points.  I've addressed them as best I can, with explanations below.

There are no boards that have this sensor natively, which is probably why there wasn't a sample before.  I use a breakout board.  I've put together something that does a build-only test for it, with support files that also serve as examples.  It may be possible to use a fixture but I couldn't figure out how to do it quickly.

I completely agree on sample dts_fixup bogosity, saving that that's how it's commonly being done now.  I've replaced it with the pattern that I've used in subsequent patches (not yet submitted as they depend on an unreviewed I2C PR).  I believe that approach provides a new user with a better clue as to what needs to be done.  If this pattern seems better then it would be nice to require it for all future samples, as consistently demonstrating best practices can reduce support requests....

Let me know your thoughts.  If you don't like either of those innovations, more detailed direction on what you'd like instead would be appreciated.

MaureenHelm replied:
You can add a dts overlay in the sample for the nrf51_ble400 board, similar to how `samples/sensor/ms5837/` defines an overlay for the nrf52840_pca10056 board.

If there is a an arduino shield board available with this sensor, a more generic solution would be to define a shield overlay like `boards/shields/x_nucleo_iks01a1/x_nucleo_iks01a1.overlay`.

pabigot replied:
Yes, that's exactly what I did to make it compile: [nrf51_ble400.overlay](https://github.com/zephyrproject-rtos/zephyr/pull/11545/files#diff-b646cfa4611ed59bbd612054e5a7ed2b).  What I'd like to do with #11737 is add an alias to that, then get rid of the dts_fixup entirely.

Mierunski replied:
Why not use aliases in existing form? I used them for I2C eeprom slave #9943 as adding something that isn't bound to any board is hard with fixups. It would be great to create new drivers that use aliases not dts_fixups

pabigot replied:
This isn't a new driver: it was added in 2016.  I'm new here and trying to figure out how much pushback people have when existing names that are visible to developers or users change.

What you suggest is exactly what I would love to do, as suggested in #11737.  Approval there seems to be blocked by big plans for an alternative solution, and disagreement on whether it's a proper use of aliases in device tree files.

What's here works and is consistent with current state-of-the-practice in Zephyr.  I'm happy to rework this if there's somebody who can make a decision on what's acceptable (I note nobody's yet been assigned this PR, though @mhelm has provided valuable feedback).  Absent that, my preference is to follow existing practice until alternatives have community consensus.

MaureenHelm commented on samples/sensor/sht3xd/src/main.c:
```diff
@@ -0,0 +1,89 @@
+
+void main(void)
+{
+	struct device *dev = device_get_binding("SHT3XD");
```
Replace the hard-coded string with `DT_SHT3XD_NAME`

pabigot commented:
The revised dts_fixup will fail checkpatch because the lines are too long.  Per a slack exchange on 2018-11-24 this is routinely accepted as an exception when noted in the PR.

pabigot commented:
Marked as do-not-merge until #11737 gets some discussion, as we might be able to get rid of the whole dts_fixup issue.

MaureenHelm commented on drivers/sensor/sht3xd/sht3xd_trigger.c:
```diff
@@ -195,7 +195,8 @@ int sht3xd_init_interrupt(struct device *dev)
 	gpio_pin_configure(drv_data->gpio, DT_SHT3XD_GPIO_ALERT_PIN,
-			   GPIO_DIR_IN | GPIO_INT | GPIO_INT_LEVEL |
+			   GPIO_DIR_IN | GPIO_INT |
+			   GPIO_INT_EDGE | GPIO_INT_DOUBLE_EDGE |
```
This change should be in the previous commit

pabigot replied:
I'd really rather not.  The sole purpose of the previous commit is to change the spelling of GPIO resource identifiers.  This changes the way the signal is generated, is required because of how the example works, and is essentially a behavior/functional change.  In practice I don't think the level interrupt could ever have worked because the application can't clear it, but nonetheless this change should either be in its own commit documenting why it was made (which I'm willing to do) or as part of the commit that motivated it (where it's documented in the commit comment).

MaureenHelm replied:
Agree, my mistake. I see a lot of pull requests that add a new sensor driver in one commit, then add a sample for that sensor in the next commit.

Let's make this change in its own commit.

pabigot replied:
done

MaureenHelm commented on samples/sensor/sht3xd/dts_fixup.h:
```diff
@@ -0,0 +1,28 @@
+
+#define DT_SHT3XD_NAME DT_NORDIC_NRF_I2C_40003000_SENSIRION_SHT3XD_44_LABEL
+#define DT_SHT3XD_I2C_ADDR DT_NORDIC_NRF_I2C_40003000_SENSIRION_SHT3XD_44_BASE_ADDRESS
+#define DT_SHT3XD_I2C_MASTER_DEV_NAME DT_NORDIC_NRF_I2C_40003000_SENSIRION_SHT3XD_44_BUS_NAME
```
Please add tabs to align:

```c
#define DT_SHT3XD_NAME			DT_NORDIC_NRF_I2C_40003000_SENSIRION_SHT3XD_44_LABEL
#define DT_SHT3XD_I2C_ADDR		DT_NORDIC_NRF_I2C_40003000_SENSIRION_SHT3XD_44_BASE_ADDRESS
#define DT_SHT3XD_I2C_MASTER_DEV_NAME	DT_NORDIC_NRF_I2C_40003000_SENSIRION_SHT3XD_44_BUS_NAME
```

pabigot replied:
done.

I believe all requested changes have been made.  Since #11737 doesn't seem likely to happen soon I have no other changes I intend to make.

MaureenHelm commented on samples/sensor/sht3xd/prj_nrf.conf:
```diff
@@ -0,0 +1,14 @@
+CONFIG_I2C=y
+CONFIG_HAS_DTS_I2C=y
+CONFIG_I2C_NRFX=y
+CONFIG_I2C_0=y
```
I'm not sure if this is also true on nordic parts, but on nxp parts you don't have to set the three configs above. If you set CONFIG_I2C=y, then the soc configures the appropriate i2c driver and the board configures the appropriate i2c instance. @carlesc, do you guys do this too?

If this is the same on nordic parts, then you can eliminate `prj_nrf.conf` entirely.

pabigot replied:
The default `prj.conf` results in errors about: 'drivers__i2c' was created without source files
I can and will get rid of `CONFIG_HAS_DTS_I2C` because that comes with the board, but the other two appear to be necessary.

carlescufi replied:
@MaureenHelm I think @Mierunski and @anangl should answer this question

pabigot replied:
I've removed all the optional pieces.  If there are changes necessary to make this file unnecessary they're probably in board files, which I propose should be a separate PR.

MaureenHelm replied:
> The default `prj.conf` results in errors about: 'drivers__i2c' was created without source files

Ok, then it looks like the nordic parts don't handle this like I'd hoped. @Mierunski, @anangl you might want to consider doing this because it allows for fewer soc-specific .conf files in the sample and test applications.

https://github.com/zephyrproject-rtos/zephyr/blob/b5ae7da2697b6e38180856f014cfe6bbc652e4e6/soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk64f12#L46-L51

https://github.com/zephyrproject-rtos/zephyr/blob/b5ae7da2697b6e38180856f014cfe6bbc652e4e6/boards/arm/frdm_k64f/Kconfig.defconfig#L73-L81

pabigot committed:
drivers: sensor: sht3xd: change from level to double edge trigger

The sensor signal cannot be cleared by the application, so enabling
level triggers causes the system to hang as the callback is invoked
repeatedly.  We want notification when the alarm state is entered, and
when it's exited, so use a double edge.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
Removed the summary line `[DNM]` as it seems unlikely that we can get rid of the dts_fixup files quickly.

MaureenHelm commented:
recheck

MaureenHelm commented:
@pabigot Could you please rebase to resolve the merge conflict?
